toponetx/utils/normalization.py

```diff
@@ -0,0 +1,383 @@
+"""Normalize of adjacency, boundary matrices of complexes."""
+
+import numpy as np
+import scipy.sparse.linalg as spl
+from numpy import ndarray
+from numpy.linalg import pinv
+from scipy.sparse import csr_matrix, diags, identity
+
+__all__ = [
+    "compute_laplacian_normalized_matrix",
+    "compute_x_laplacian_normalized_matrix",
+    "compute_kipf_adjacency_normalized_matrix",
+    "compute_xu_asymmetric_normalized_matrix",
+    "compute_bunch_normalized_matrices",
+    "_compute_B1_normalized_matrix",
+    "_compute_B1T_normalized_matrix",
+    "_compute_B2_normalized_matrix",
+    "_compute_B2T_normalized_matrix",
+    "_compute_D1",
+    "_compute_D2",
+    "_compute_D3",
+    "_compute_D5",
+]
+
+
+def compute_laplacian_normalized_matrix(L):
+    """
+    Normalize the Laplacian matrix.
+
+    Parameters
+    ----------
+    L : scipy.sparse.csr_matrix
+        The Laplacian matrix.
+
+    Returns
+    -------
+    scipy.sparse.csr_matrix
+        The normalized Laplacian matrix.
+
+    Notes
+    -----
+    This function normalizes the Laplacian matrix by dividing it by the largest eigenvalue.
+    """
+    topeigen_val = spl.eigsh(L.asfptype(), k=1, which="LM", return_eigenvectors=False)[
+        0
+    ]
+    return L * (1.0 / topeigen_val)
+
+
+def compute_x_laplacian_normalized_matrix(L, Lx):
+    """
+    Normalize the up or down Laplacians.
+
+    Parameters
+    ----------
+    L : scipy.sparse.csr_matrix
+        The Laplacian matrix.
+    Lx : scipy.sparse.csr_matrix
+        The up or down Laplacian matrix.
+
+    Returns
+    -------
+    scipy.sparse.csr_matrix
+        The normalized up or down Laplacian matrix.
+
+    Notes
+    -----
+    This function normalizes the up or down Laplacian matrices by dividing them
+    by the largest eigenvalue of the Laplacian matrix.
+    """
+    assert L.shape[0] == L.shape[1]
+    topeig = spl.eigsh(L.asfptype(), k=1, which="LM", return_eigenvectors=False)[0]
+    return Lx * (1.0 / topeig)
+
+
+def compute_kipf_adjacency_normalized_matrix(
+    A_opt, add_identity=False, identity_multiplier=1.0
+):
+    """
+    Normalize the adjacency matrix using Kipf's normalization.
+
+    Typically used to normalize adjacency matrices.
+
+    Parameters
+    ----------
+    A_opt : numpy.ndarray
+        The adjacency matrix.
+    add_identity : bool, optional
```

Comment:
```suggestion
    add_identity : bool, default=False
```
In general optional boolean flags feel weird, there is nothing optional about them. They default to true or to false, that's it.